Add proposal for supporting SSL for KafkaBridge (#171)
strimzi/proposals: 1xx-enable-ssl-for-kafka-bridge.md

## Proposal

This proposal adds new configurations to enable SSL server and specify locations of certificate and key files in PEM format so that client connections to HTTP Bridge are encrypted. The new configurations will be added to the current `http` prefixed configurations:
- http.ssl.enable
- http.ssl.port
- http.ssl.keystore.location
- http.ssl.keystore.key.location
- http.ssl.enabled.protocols
- http.ssl.enabled.cipher.suites

When `http.ssl.enable` is set to true, HTTP Bridge server will be started with SSL enabled and the key and certificate loaded from the locations defined by `http.ssl.keystore.location` and `http.ssl.keystore.key.location` configurations. If `http.ssl.port` configuration is not set, the server port will be set to `8443` by default. This will allow users to make TLS encrypted connections from their clients for all the HTTP Bridge endpoints for client operations. However, connections to the following endpoints for monitoring will stay unencrypted:
- `/healthy`
- `/ready`
- `/metrics`
- `/openapi`
- `/openapiv2`
- `/openapiv3`

> - `/info`

Reviewer: Maybe it is worth explaining why should these not be on the SSL listener as well. I guess I can understand that for metrics and possibly the ready / healthy endpoints. Not so sure about the OpenAPI endpoints and the /info endpoint. Wouldn't they be useful for clients as well?

Reviewer 2: Currently `/info` provides only bridge version, while the OpenAPI endpoints provide the full OpenAPI YAML definition which is useful in case of integration with API management systems like 3scale (a use case I worked in the past) for dynamic discovery of the operations. Having them behind the SSL could make such integration harder (or break it) if the other side doesn't allow SSL connection first.

Reviewer: I don't know TBH. That was how it started before I started digging into why are we removing these from 8080/8443. As far as I'm concerned, we can do just healthcecks and metrics.

> When `http.ssl.enable` is set, these endpoints will have a separate listener, that is always PLAINTEXT. It will use the port defined by the existing `http.port` that is set to `8080` by default. The purpose of these endpoints is to be exposed only internally for the health checks, getting OpenAPI specification and metrics collection. In the future, we are likely to implement authentication and authorization for HTTP Bridge. By having separate listeners, these monitoring endpoints will not be affected by such changes.

Reviewer: As I said before, this just postpones the issue for when authentication is introduced. I think the proposal should be more forward-thinking and use a completely separate port for the metrics and healthchecks.

`http.ssl.enabled.cipher.suites` can be configured with a comma separated list of cipher suites that server will support. If not set, Vertx uses the default cipher suites provided by the underlying JDK SSL/TLS engine.

### HTTP Bridge managed by Strimzi operator

When deploying KafkaBridge custom resources with the Strimzi operator, users can set these configurations via `spec.http` in their KafkaBridge CR. Users can enable SSL by setting the new field `spec.http.sslEnable` which is set to false by default. When it is set to true, users have to provide the certificate and key via Kubernetes secret that is referenced in the `certificateAndKey` field.

Let's take a look at an example of KafkaBridge CR with the new fields:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092

>   http:

Reviewer: Why not have a separate section `https`?

Reviewer: Why will we use `http.port`? Why not allow using both HTTP and HTTPs in parallel?

Reviewer 2: > But they are odd ones to be configured under http prefix when running a standalone bridge. I think they should have been under kafka prefix because will we need to enable consumer or producer specifically only for HTTP or HTTPS?

The `kafka.` prefix is used for all configuration related to the internal Kafka consumer, producer and admin. Any configuration from there, is just used to configure the corresponding Kafka clients.
The `http.consumer.enabled` and `http.producer.enabled` represents the HTTP part of them. I remember it was a community ask to be able to disable consumer or producer HTTP endpoint (in order to get `503` `Service Unavailable`) if not enabled.

> If we allow running both, will there users who would want to run only HTTPS ? If so we would need to have an option to disable HTTP.

I would agree with Jakub. I think for security concerns we'll have for sure people enabling HTTPS and ... disabling HTTP (which is actually not supported). What I am not sure is if people we'll enable them both at the same time.
I think that compared with the Kafka listeners we had more problems because we have much more integration within Kubernetes, so you can have Ingress, Routes, etc etc ...
For the bridge, it can run as standalone as well, so I don't think we can have more than HTTP and HTTPS. Everything you want to expose it outside of Kubernetes is up to you, even the operator doesn't provide anything (just an internal Service).

## Proposal

This proposal adds new configurations to enable SSL server and specify locations of certificate and key files in PEM format so that client connections to HTTP Bridge are encrypted. The new configurations will be added to the current `http` prefixed configurations:
- http.ssl.enable
- http.ssl.port
- http.ssl.keystore.location
- http.ssl.keystore.key.location
- http.ssl.protocol
- http.ssl.enabled.protocols
- http.ssl.add.cipher.suite
- http.ssl.remove.cipher.suite

When `http.ssl.enable` is set to true, HTTP Bridge server will be started with SSL enabled and the key and certificate loaded from the locations defined by `http.ssl.keystore.location` and `http.ssl.keystore.key.location` configurations. If `http.ssl.port` configuration is not set, the server port will be set to `8443` by default. This will allow users to make TLS encrypted connections from their clients for all the HTTP Bridge endpoints for client operations. However, connections to the following endpoints for monitoring will stay unencrypted:
- `/healthy`
- `/ready`
- `/metrics`
- `/openapi`
- `/openapiv2`
- `/openapiv3`

> - `/info`

Reviewer 2: I think with option 2 we have an opportunity to separate administrative endpoints from operational ones. Even scraping metrics on the same HTTP interface where you are sending and receiving messages could have an impact on performance. Separating the HTTP listeners could be better (and this is even not related to the TLS addition). Anyway let's see what others think.

When `http.ssl.enable` is set, these endpoints will have a separate listener, that is always PLAINTEXT. It will use the port defined by the existing `http.port` that is set to `8080` by default. The purpose of these endpoints is to be exposed only internally for the health checks, getting OpenAPI specification and metrics collection. In the future, we are likely to implement authentication and authorization for HTTP Bridge. By having separate listeners, these monitoring endpoints will not be affected by such changes.

> If `http.ssl.protocol` is configured, the protocol will be added to the current or default list of enabled protocols. If `http.ssl.enabled.protocols` is configured with a comma separated list, it will replace the default list of enabled secure transport protocols that Vertx sets.

Author: From reading the [Vertx Java doc](https://vertx.io/docs/apidocs/io/vertx/core/net/SSLOptions.html), I don't think these 2 methods are used differently. It is either used to overwrite the default list of protocols or append a protocol to the ordered list of protocols Vertx server would accept. This means we cannot mimic `ssl.protocol` of Kafka, so I think your suggestion of having just `http.ssl.enabled.protocols` makes sense. 

[`"TLSv1", "TLSv1.1", "TLSv1.2", "TLSv1.3"`](https://vertx.io/docs/apidocs/io/vertx/core/net/SSLOptions.html#DEFAULT_ENABLED_SECURE_TRANSPORT_PROTOCOLS) is the default list of protocols Vertx uses. Do we need to set this to the default of Kafka by default?

> Vertx uses the default cipher suites provided by the underlying JDK SSL/TLS engine and allows users to append a cipher suite to the ordered suites or remove one from it. HTTP Bridge users can use the new configurations `http.ssl.add.cipher.suite` and `http.ssl.remove.cipher.suite` to add or remove a cipher suite from the default list of enabled cipher suites.

Reviewer 2: Like for the SSL protocols, we can try to have something similar to Kafka with just `ssl.cipher.suites` which is a comma separated list and if not set it defaults to the available ones (which has to be seen what it exactly means for Vert.x).?

> Vertx uses the default cipher suites provided by the underlying JDK SSL/TLS engine and allows users to append a cipher suite to the ordered suites or remove one from it. HTTP Bridge users can use the new configurations `http.ssl.add.cipher.suite` and `http.ssl.remove.cipher.suite` to add or remove a cipher suite from the default list of enabled cipher suites.

Author: I think that would make sense. There is no bulk setter for ciphers suites, but we can take a list of cipher suites, loop through them with addEnabledCipherSuite() method. Now thinking about it, `removeEnabledCipherSuite` is probably not that useful, as it's something we set up before starting the server. 

>  (which has to be seen what it exactly means for Vert.x)?
It seems like Vertx does not set any defaults so probably defaults provided by the JDK is used.

## Motivation

The connection between HTTP Bridge and Kafka cluster can be secured however, currently there is no support for securing the connections to HTTP Bridge from clients. The current recommendation is to use firewalls or API gateways to secure the client connections. This requires users to make additional configurations and does not provide the best user experience. 

Moreover, some of the endpoints are for internal use only, such as `/healthy`, `/ready` and `/metrics` but they are exposed on the same listener and the same port as the other endpoints that are used by external clients. This means the endpoints for internal use only are also exposed to external clients. And if SSL is enabled, it will introduce more operational complexity as these internal endpoints would need to be configured with TLS and certificates.

## Proposal
This proposal also adds new configurations to enable SSL server and specify locations of certificate and key files in PEM format so that client connections to HTTP Bridge can be encrypted. The new configurations will be added to the current `http` prefixed configurations:
- http.ssl.enable
- http.ssl.keystore.location

> - http.ssl.keystore.key.location

Reviewer: What is the difference between `http.ssl.keystore.location` and `http.ssl.keystore.key.location`? Also, should we:
* Have some type field? What formats will we support? You seem to mention PEM only. But given how common JKS/PKCS12 is in Java land, we should be at least ready to support them. They would also need a password field I guess.
* Have some field to pass the keystore public and private key directly? E.g. `http.ssl.keystore.key` and `http.ssl.keystore.certificate.chain`?
* Would it make more sense to have the options named as `https`? E.g. `https.enable`?

- http.ssl.enabled.protocols
- http.ssl.enabled.cipher.suites

> - http.admin.port

Reviewer: It does not really have any admin options. Should we try to find some better name for it?

Reviewer: Probably `http.internal.port` or `http.healthmetrics.port` makes most sense. @ppatierno WDYT?

Reviewer 2: I see `http.healthmetrics.port` too much specific. I liked the `management` one but I guess it's going to be refused because too similar to `admin`. I would go with `internal` at this point ... even if it's not so clear what actually it provides (it just says, it's internal).

Reviewer: I get the ugliness of all the alternatives we have here.  But do we have really anything better? I mean it does not do any management, or? Would it do some management one day?

Reviewer 2: I was taking a look around for other project and it looks like Spring and Quarkus are using the `management` term for this kind of thing. AFAIU Quarkus has a `management` port exposing health, metrics, info and openapi (very similar to the ones we have) https://quarkus.io/guides/management-interface-reference and https://www.baeldung.com/quarkus-management-interface
Spring seems to expose much more still through a `management` port.
At this point, could this name be considered quite common if used by this very well know development framework? Also should we revisit the presence of /info and /openapi on it?

Reviewer: Ok, I guess I can live with it if others use the same bad name as well :-). Maybe one day we will add some management to it, some dynamic reconfiguriation or something 🤷

The rest of the endpoints will stay on the existing listener with the default port 8080 or 443 if SSL is enabled. 

This brings a clear separation of internal and external use for the endpoints and makes it simpler to enable SSL for the listener that is used by external clients. It also brings a better performance and security because admin type of operations such as collecting metrics are isolated from traffic for Kafka operations.

### HTTP Bridge managed by Strimzi operator

When deploying KafkaBridge custom resources with the Strimzi operator, users can set these configurations via `spec.http` in their KafkaBridge CR. Users can enable SSL by setting the new field `spec.http.sslEnable` which is set to false by default. When it is set to true, users have to provide the certificate and key via Kubernetes secret that is referenced in the `certificateAndKey` field.

Let's take a look at an example of KafkaBridge CR with the new fields to enable SSL:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092
  http:
    port: 8443

>     adminPort: 8081 #NEW FIELD

Reviewer: Should this be really configurable? I would probably argue against it. I don't think similar ports are configurable in other operands.

    sslEnable: true #NEW FIELD
    certificateAndKey: #NEW FIELD
      secretName: my-secret
      certificate: public.crt
      key: private.key
...
```

> The liveness and readiness probes will connect to the port defined by `adminPort` (or 8080 if not set).  

Reviewer: Should this be 8081 instead as that is the default for the admin port?

Author: sorry my mistake, it is 8081.

Let's take a look at an example of KafkaBridge CR with the new fields to enable SSL:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092
  http:
    port: 8443
    sslEnable: true #NEW FIELD
    certificateAndKey: #NEW FIELD
      secretName: my-secret
      certificate: public.crt
      key: private.key
...
```
The liveness and readiness probes will connect to port 8081 (instead `http.port` or 8080). If users have metrics set up, they would need to use port 8081 to connect to the `metrics` endpoint and cannot reconfigure this port number as it's only for internal use.

An example of KafkaBridge CR with other possible SSL configurations:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092
  http:
    port: 8443

>     sslEnable: true 

Reviewer 3: Yeah that's something what I was thinking about as well.

      serverCertChainAndKey: #NEW FIELD
        secretName: my-secret
        certificate: public.crt
        key: private.key
      config: #NEW FIELD
        ssl.enabled.protocols: "TLSv1.1, "TLSv1.2", "TLSv1.3"
        ssl.enabled.cipher.suites: "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
...
```

## Future Improvements

### Support more formats for the keystore

> We can support formats such as `PKSC12` and `JKS` as they are common for Java applications. If we support these formats, then we would need to add the following configurations:

Reviewer 4: ```suggestion
We can support formats such as `PKCS12` and `JKS` as they are common for Java applications. If we support these formats, then we would need to add the following configurations:
```

- http.ssl.keystore.type
- http.ssl.keystore.password

### Support mTLS

> In the future, we can support mutual TLS for the client connections to HTTP Bridge as well. This could be implemented by adding a field `trustedCertificates`, for example:

Reviewer 4: ```suggestion
In the future, we can support mutual TLS for the client connections to HTTP Bridge as well. This could be implemented by adding a new `trustedCertificates` property in the `tls` section, for example:
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092
  http:
    port: 8443
    tls: 
      serverCertChainAndKey: 
        secretName: my-secret
        certificate: public.crt
        key: private.key
      trustedCertificates: #NEW FIELD
        - secretName: my-cluster-cluster-cert
          pattern: "*.crt"
        - secretName: my-cluster-cluster-cert
          pattern: "*.crt"    
...
```

> This is not in the scope of this proposal, as we will likely to support multiple authentication mechanisms in the future.

Reviewer 4: ```suggestion
This is not in the scope of this proposal, but we are likely to support multiple authentication mechanisms in the future.
```

> In the future, we are also likely to implement authentication and authorization for HTTP Bridge. By having separate listeners, the internal endpoints will not be affected by such changes and will make it easier to implement them.

Reviewer 4: ```suggestion
We also plan to implement authentication and authorization for the HTTP Bridge. By using separate listeners, internal endpoints are isolated from these changes, making implementation simpler and less disruptive.
```

### Automating certificate management

> Users can choose to allow the operator to manage the certificate instead of providing their own certificate and key. This can be implemented by adding another boolean field on the KafkaBridge CR such as `generateCertificate`. The operator then could issue the certificate using the same CA for their Kafka cluster as we do today for internal components such as Cruise Control and Kafka Exporter, however with the upcoming support for cert-manager, the way the operator manages certificates will change. This is the reason we are leaving this feature for future improvement.

Reviewer 4: ```suggestion
Users can choose to allow the operator to manage the certificate instead of providing their own certificate and key. This can be implemented by adding another boolean field on the `KafkaBridge` CR such as `generateCertificate`. The operator could issue certificates using the same CA for their Kafka cluster, as is currently done for internal components like Cruise Control and Kafka Exporter. However, with upcoming support for cert-manager, certificate management is expected to change. For this reason, we are deferring this feature for implementation in the future.
```

## Affected/not affected projects

- strimzi-kafka-bridge
- strimzi-cluster-operator

## Compatibility

> This will introduce a breaking change for users who collect metrics on port 8080 or the port defined by `http.port` configuration. Also users who run Bridge on baremetals and do their own manual healtchecks using `/healthy` and `/ready` endpoints will be affected as well. These users have to reconfigure their healthcheck and monitoring system to use port 8081 or reconfigure their Bridge server with `http.management.port`.

Reviewer 4: ```suggestion
This will introduce a breaking change for users who collect metrics on port 8080 or the port defined by `http.port` configuration. Also users who run the HTTP Bridge on bare metal and do their own manual health checks using `/healthy` and `/ready` endpoints will be affected as well. These users have to reconfigure their health check and monitoring system to use port 8081 or reconfigure their Bridge server with `http.management.port`.
```

## Rejected alternatives

### Automating certificate management as part of this proposal

> - Currently the operator does not manage certificates for other custom resources, as they are designed to be independent, it only manages certificates for its internal components such as Cruise Control and Kafka Exporter, that are deployed as part of the Kafka custom resource. This will introduce a new behaviour to the operator and would require taking other custom resources such as KafkaConnect into consideration as well, because we would potentially need to manage certificates for it too. We are currently in the process of refactoring how the operator manages certificates to support external certificate manager as proposed [here](https://github.com/strimzi/proposals/pull/135), therefore it is not the right time to introduce such new behaviour, as it would not be future proof. Instead, it was decided to leave it for future improvement. 

Reviewer 4: ```suggestion
- Currently, the Cluster Operator does not manage certificates for other custom resources, as they are designed to operate independently. It only handles certificates for internal components, such as Cruise Control and Kafka Exporter, which are deployed as part of the `Kafka` custom resource. Extending certificate management to other resources, like `KafkaConnect`, would introduce new behavior and require broader changes. Since we’re looking at refactoring certificate handling to support external managers (see the proposal for the [External Certificate Manager](https://github.com/strimzi/proposals/pull/135)), now is not the right time to introduce this change. To ensure future compatibility, we’ve deferred it as a potential improvement.
```

> - Continue using a single interface and only use separate listeners when SSL is enabled so that we don't introduce a breaking change to existing users. Given that we will be introducing the first major version of Bridge, introducing a breaking change is acceptable and it makes much more sense to have separate listeners for better separation of concerns regardless of TLS. In the future, we are likely to implement authentication and authorization for HTTP Bridge. By having separate listeners, th admin endpoints will not be affected and implementation will be simpler. 

Reviewer 4: ```suggestion
- Continue using a single interface and only use separate listeners when SSL is enabled so that we don't introduce a breaking change to existing users. Given that we will be introducing the first major version of Bridge, introducing a breaking change is acceptable and it makes much more sense to have separate listeners for better separation of concerns, , whether or not TLS is enabled. In the future, we are likely to implement authentication and authorization for HTTP Bridge. By having separate listeners, the admin endpoints will not be affected and implementation will be simpler. 
```

## Current situation

Strimzi does not support TLS/SSL for the HTTP Bridge, so client connections to the HTTP Bridge are unencrypted. A single listener serves all endpoints for Kafka operations and admin operations such as health checks and monitoring. The default port is 8080.

## Motivation

The connection between HTTP Bridge and Kafka cluster can be secured. However, there is currently no support for securing client connections to the HTTP Bridge. The current recommendation is to use firewalls or API gateways to secure the client connections. This requires users to make additional configurations and does not provide the best user experience. 

Moreover, some of the endpoints are for internal use only, such as `/healthy`, `/ready` and `/metrics` but they are exposed on the same listener and the same port as the other endpoints that are used by external clients. This means the endpoints for internal use only are also exposed to external clients. And if SSL is enabled, it will introduce more operational complexity as these internal endpoints would need to be configured with TLS and certificates.

## Proposal

> This proposal introduces adds new configurations to enable an SSL server and load a keystore certificate and key in PEM format so that client connections to HTTP Bridge can be encrypted. The new configurations will be added to the existing `http` prefixed configurations:

Reviewer 5: ```suggestion
This proposal introduces new configurations to enable an SSL server and load a keystore certificate and key in PEM format so that client connections to HTTP Bridge can be encrypted. The new configurations will be added to the existing `http` prefixed configurations:
```

> - http.ssl.enable

Reviewer 5: Just out of curiosity, adding the `https` instead of `http.ssl` prefix wouldn't be better?

- http.ssl.keystore.certificate.location
- http.ssl.keystore.key.location
- http.ssl.keystore.certificate.chain
- http.ssl.keystore.key
- http.ssl.enabled.protocols
- http.ssl.enabled.cipher.suites
- http.management.port

> When `http.ssl.enable` is set to `true`, the HTTP Bridge server starts with SSL enabled and the new configurations will allow users to define locations of the keystore files or certificate chain and key in `PEM` format. Additional formats such as `PKCS12` and `JKS` might be supported in the future if users request them.

Reviewer 2: So AFAIU the user should set `http.ssl.keystore.certificate.location` together with `http.ssl.keystore.key.location` OR `http.ssl.keystore.certificate.chain` together with `http.ssl.keystore.key`. Is my understanding right? Should we make it more explicit?

`http.ssl.enabled.protocols` can be configured with a comma separated list of enabled secure transport protocols that the server will accept from connecting clients. If not set by the user, the server will use `TLSv1.2,TLSv1.3` as the default. 

> `http.ssl.enabled.cipher.suites` can be configured with a comma separated list of cipher suites that the server will support. If not set, the default list of cipher suites provided by the underlying JDK SSL/TLS engine will be used.

Reviewer 2: Same questions as for the protocols.

> If the existing `http.port` is not set and `http.ssl.enable` is set to `true`, the server port will be set to `443` by default. This will allow users to make TLS encrypted connections from their clients for all the HTTP Bridge endpoints for client operations.

Author: This was also discussed early on, I will find and shared the link here. When running bridge with the operator, we actually enforce `8443` so maybe it is better to make it `8443` for standalone as well?

> If the existing `http.port` is not set and `http.ssl.enable` is set to `true`, the server port will be set to `443` by default. This will allow users to make TLS encrypted connections from their clients for all the HTTP Bridge endpoints for client operations.

Author: Initially the proposal had 8443, but I remember changing it based on a discussion with @ppatierno but can't remember or find it. I think we had decided that 443 was more standard HTTPS port?

This proposal also adds support for a separate listener for internal endpoints such as:
- `/healthy`
- `/ready`
- `/metrics`

> The internal endpoints will use the port defined by the new configuration property `http.management.port`, defaulting to 8081 if unspecified. These endpoints will no longer be accessible via the default listener port defined by `http.port` or port 8080. All other endpoints will continue to use the existing listener on port 8080, or 443 if SSL is enabled.

Reviewer 2: Did you try what happens if ... the user DOESN'T enable TLS so they get default HTTP port 8080 but then they also set `http.management.port` to 8080? If you start two servers in the bridge, it's not going to work because the first one will keep the port and the other one will find the port already used. At the same time, as user, I would expect this to work. I just want operations and admin on the same 8080 port as today. Have you taken into account this scenario in terms of implementation?

### HTTP Bridge managed by Strimzi operator

When deploying `KafkaBridge` custom resources with the Strimzi operator, users can set these configurations in the `spec.http` section of their `KafkaBridge` CR. Users can enable SSL by setting the new field `spec.http.sslEnable`, which is set to `false` by default. When it is set to `true`, users have to provide the certificate and key in PEM format using a Kubernetes `Secret` that is referenced in the `certificateAndKey` field.

Let's take a look at an example of the `KafkaBridge` CR with the new fields to enable SSL:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092
  http:
    port: 8443
    tls: #NEW FIELD
      serverCertChainAndKey: #NEW FIELD
        secretName: my-secret
        certificate: public.crt
        key: private.key
...
```

> The liveness and readiness probes will connect to port 8081 (instead of `http.port` or 8080). If users have metrics set up, they would need to use port 8081 to connect to the `metrics` endpoint and cannot reconfigure this port number as it's only for internal use.

Reviewer 2: So we don't allow to change the management port within the `KafkaBridge` custom resource? Why?

# Support TLS/SSL on the HTTP interface 

This proposal adds an option to enable TLS encryption on HTTP interface so that requests from clients to the Kafka Bridge can be sent with encryption.

## Motivation

The connection between HTTP Bridge and Kafka cluster can be secured however, currently there is no support for securing the connections to HTTP Bridge from clients. The current recommendation is to use Network policies and firewalls, reverse proxies such as OAuth 2.0 or API gateways to secure the client connections. This requires users to make additional configurations but does not provide the best user experience.  

## Proposal

This proposal adds new configurations to enable SSL server and specify locations of certificate and key files in PEM format so that client connectionts to HTTP Bridge is encryped. The new configurations will be added to the current `http` prefixed configurations:
- http.ssl.enable
- http.ssl.keystore.location
- http.ssl.keystore.key.location
- http.ssl.protocol
- http.ssl.enabled.protocols

> When `http.ssl.enable` is set to true, Kafka Bridge server would be started with ssl enabled and the key and certificate loaded from the locations defined by `http.ssl.keystore.location` and `http.ssl.keystore.key.location` configurations. If `http.port` configuration is not set, the server port will be set to `8443` by default. This will allow users to make HTTPS connections from their clients, and the server would not accept unsecured connections.

Reviewer: I think this requires a bit more consideration ...
* Can we really afford to have only the TLS listener? What about healthchecks, metrics, etc.
* If we have both PLAIN and TLS listener, what about the security (not jsut TLS, but in the future also Auth and so on)?
* So maybe we need to split-of metrics and healthchecks into a separate listener in the first place?

Author: Thanks for the feedback Jakub. 
> Can we really afford to have only the TLS listener? What about healthchecks, metrics, etc.
>So maybe we need to split-of metrics and healthchecks into a separate listener in the first place?

You right, we have to consider the metrics and health endpoints for monitoring and readiness/liveness. As you suggested, we can have both PLAIN and TLS listener and then when ssl enabled, the PLAIN listener is configured for only `/healthy`, `/ready` and `/metrics` endpoints and the TLS listener is for the other endpoints. 

> If we have both PLAIN and TLS listener, what about the security (not jsut TLS, but in the future also Auth and so on)?
I was going to scope the auth part out of this proposal but perhaps should mention different auth mechanism we could support such as HTTP Basic Authentication and JWT in the future improvement. I will make an update on this.

Reviewer: > I was going to scope the auth part out of this proposal but perhaps should mention different auth mechanism we could support such as HTTP Basic Authentication and JWT in the future improvement. I will make an update on this.

Right. But even with auth being out of scope for this proposal, we need to think about it a bit. What my concern is is the following ... if we now create two interfaces -> plain and TLS with the idea that the plain one will be used for healthchecks and metrics, it does not seem to work long time. Because you will face the same issue once again:
* If you enable authentication on the plaintext interface, healcthchecks and metrics will stop working
* Having a plaintext interface always enabled for metrics and healthchecks, means you cannot enforce TLS either.

So, maybe, we need in the first lace to separate healthchecks and metrics from the regular listeners and move them to completely separate endpoint.

Reviewer 2: I agree on moving healthchecks and metrics endpoints to a different HTTP plain interface. It should be the same for the `/openapi` endpoint as well. Said that if we are allowing two different interfaces, PLAIN and TLS, we would also need an additional `http.ssl.port` (other than the `http.port`). Also ... should we make the port related to the healthchecks, metrics and openapi endpoint being configurable as well? If we hard code such port, it could be a problem for users who run bridge on bare metal/VMs and that port is already owned by something else. Maybe an additional `http.management.port` (or something similar better than `management`? But it seems to be used in other frameworks, i.e. SpringBoot AFAIU). Of course it's going to have an impact on the operator as well to configure the additional port and the Kube probes.

Author: I have added `http.ssl.port` but left the management endpoints to carry on using the http.port so that it doesn't have an impact on kube probes and compatibility issue.

> If `http.ssl.protocol` is configured, the protocol will be added to the current or default list of enabled protocols. If `http.ssl.enabled.protocols` is configured, the list will replace the default list of enabled secure transport protocols that Vertx sets.

Reviewer: Aren't there some existing Vert.x options to configure this? If not, do we need the same options also for cipher suites? As that is another common thing users configure.

 
When running KafkaBridge with Strimzi operator, users can choose to either provide these configurations via `spec.http` in their KafkaBridge CR or allow the operator to manage the certificate using the same CA for the Kafka cluster.

Users can enable SSL by setting the new field `spec.http.sslConfigurations`. Under this new field, they have to set `generateCertficiate` field that accepts boolean value and is set to false by default. If it is not set in the CR, that means the user chooses to provide their own certificate via Kubernetes secret, in which case, they have to set `certificateAndKey` to refer to the secret that contains the certificate and key. 

> If `generateCertificate` is set true, the operator will issue the certificate using the same CA for their Kafka cluster. When operator is reconciling KafkaBridge resource, it is not aware of which Kafka cluster the HTTP Bridge is connected to. So if the Kafka cluster is running in a different namespace, therefore the CA secrets are not in the same namespace as the KafkaBridge resource, the operator needs to know where to fetch them. In this case users have to also set `caNamespace`. However, if their Kafka cluster and KafkaBridge resources are running in the same namespace, they don't need to set this field, as the operator would look for them in the namespace of the KafkaBridge resource by default.

Reviewer: I personally would definitely avoid this.
* The resources were designed as independent and IMHO they should stay that way
* How will you handle situations when the cluster is deleted etc.?
* How will you know which cluster to pick from given namespace?
* How can you guarantee that this is _future proof_ with regards to the future Cert Manager support in the Kafka CR? Also, some of the thoughts in the past were that after/if we manage to get proper support for Cert Manager, we would be able to deprecate and remove our own CA management anyway.

Author: Good points! Tbh, I have gone back and forth on how we would or should do certificate management for another CR, since this is not something we do today. We only manage certificates for our internal components such as CC, kafka-exporter etc. So what you are suggesting is that, the users would have to provide the certAndKey and manage their own certificate, at least for now? When Cert Manager support is completed, perhaps we could see how we can use it to issue a certificate for Bridge resource in the future? This would certainly simplify the proposal more.

Reviewer 2: Also taking into account that the Kafka cluster could be even running on a totally different Kube instance, cloud, bare metal/VMs and so on (not sure about the point of doing that but it's possible).
At the same time I could expect users asking for some automation, at least when the Kafka cluster is within the same namespace of the bridge (which is most of the times?). Should maybe the bridge have an additional annotation like `strimzi.io/cluster: my-cluster` as we have on KafkaTopic and KafkaUser to make clear the cluster it's referring to. Maybe making the `bootstrapServers` field optional in such a case. This way the operator would know which CA to use to generate a cert for the bridge. But in general it depends on when ceret-manager will land, because this part of the code could become obsolete soon.

metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092
  # HTTP configuration (required)
  http:
    port: 443
    sslConfigurations:
      generateCertficiate: true
      certificateValidityDays: 90 #CA default is 365
      certificateRenewalDays:  14 #CA default is 30
...
```

Users can also configure the secure transport protocols by setting these additional fields that are optional:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092
  http:
    port: 443
    sslConfigurations:
      sslProtocol: "TLSv1.3" #This will add the protocol to the default list of enabled protocols.
...
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: my-bridge
spec:
  replicas: 3
  bootstrapServers: <cluster_name>-cluster-kafka-bootstrap:9092
  # HTTP configuration (required)
  http:
    port: 443
    sslConfigurations:
      # This will override the default list of enabled protocols.
      # As of Vertx version 5.0.3, the default list is  "TLSv1, TLSv1.1, TLSv1.2, TLSv1.3".

>       sslEnabledProtocols: "TLSv1.2,TLSv1.3" 

Reviewer: I understand we probably want dedicated API for seleting the keystore. But do we want it also for all the other options? Why not keep things such as protocol configuration for a regular `config` section?

Author: So that it can be used for Kafka clients as well?

Reviewer: No, I'm more concerned that overtime you would have lot of different fields here in the API with different options just to cover everyones requirement. Having it in ther `config` makes this easier as that is just a `Map`.
